### DISCUSSION TITLE
AdventOfCode 2020

### DISCUSSION THREAD

**Discussion by astynax** [ruHaskell/ruhaskell#112] [Show and tell]
Пробуем обсуждать [AdventOfCode](https://adventofcode.com/2020) тут, чтобы не оффтопить в чатах.

**Comment by astynax** [comment]
В №16 наконец-то интересная задачка для парсинга - нужно правильно ограничить жадность, вот это вот

**Reply by ShrykeWindgrace** [reply]
Мне показалось, что наоборот - скучный парсинг. Очередная комбинация `splitOn` и `words`, ничего нового по сравнению с предыдущими днями.

**Reply by astynax** [reply]
Я имел в виду парсинг с помощью парсер-комбинаторов. Поэтому написал про "жадность".

**Reply by aragaer** [reply]
А я не придумал как парсер-комбинаторами разбить на части, поэтому использовал их только для отдельных строк.

**Reply by astynax** [reply]
@aragaer стоило добить. Я потому про интересность и написал - это отличное упражнение.

Разделитель здесь - пустая строка. Нужно так описать "много того-то", чтобы пустая строка не матчилась.

**Reply by aragaer** [reply]
Впрочем, переписал. Единственное место, где это понадобилось, это первый кусок.
```haskell
(rule <* char '\n') `manyTill` char '\n'
```

**Comment by aragaer** [comment]
В 17 задаче хочется как-то обобщить две части. Я пока придумал только один способ - использую оба раза 4 измерения, но в первой части при поиске "потенциальных новых активируемых" для 4 размерности всегда беру значение 0.
```haskell
inactive3 :: Set Point -> Set Point
inactive3 r = fromList [Point x y z 0 | x <- f x, y <- f y, z <- f z]
  where f g = [findMin (S.map g r) - 1..findMax (S.map g r) + 1]

inactive4 :: Set Point -> Set Point
inactive4 r = fromList [Point x y z w | x <- f x, y <- f y, z <- f z, w <- f w]
  where f g = [findMin (S.map g r) - 1..findMax (S.map g r) + 1]
```

**Reply by astynax** [reply]
Я обобщил для произвольной размерности :) Аннотируешь тип и получаешь требуемую:
```haskell
solve :: Set (() :*: Int :*: Int) -> (Int, Int)
solve w =
  ( Set.size $ 6 `times` step @(() :*: Int :*: Int :*: Int)         $ expand w  -- 3D
  , Set.size $ 6 `times` step @(() :*: Int :*: Int :*: Int :*: Int) $ expand w  -- 4D
  )
```

**Comment by ShrykeWindgrace** [comment]
А вот сегодня парсер как раз поинтереснее будет=)

**Reply by astynax** [reply]
Чем это? Разбор арифметических выражений - рутина и скукота.

**Reply by aragaer** [reply]
Ну да, в документации parsec практически готовый пример.

**Comment by aragaer** [comment]
19 день через "функции" получился, а вот чтобы из набора правил сделать парсер - не очень.

```haskell
validate :: Ruleset -> [String] -> [String]
validate ruleset strings = filter (any (=="") . match ruleset (Sequence [0])) strings

match rset (Character c) "" = fail "no match"
match rset (Character c) (x:xs) | c == x    = pure xs
                                | otherwise = fail "no match"

match rset (Sequence []) s     = pure s
match rset (Sequence (r:rs)) s = do
  r <- rset ! r
  m <- match rset r s
  match rset (Sequence rs) m
```
```haskell
applyRuleset rs = parse (parserFromRule rs (Sequence [0]) >> eof) ""

parserFromRule rset (Character c) = char c >> pure ()
parserFromRule rset (Sequence s) = mapM_ (fromChoice . (rset !)) s
  where fromChoice = choice . map (try . parserFromRule rset)
```
Первый вариант работает корректно, а во втором как будто не хватает откатывания после жадного захвата.

**Comment by astynax** [comment]
У меня день 19 в целом получился, но пришлось ограничить перебор вариантов для циклов :( Даже после того, как сделал его нежадным, всё равно где-то зацикливается, а времени докапываться до сути нет...

**Reply by aragaer** [reply]
Странно, ведь там все правила имеют ширину строго больше нуля. То есть каждое применение правила уменьшает длину оставшейся строки.

**Comment by astynax** [comment]
У меня архитектура немного не туда ушла — я просто перебирал варианты с
возрастающей длиной. А так, да, можно было отслеживать уменьшение инпута

**Comment by astynax** [comment]
Чёт у меня ресурсов не хватает у ноута, чтобы day22 досчитать %)

**Reply by ShrykeWindgrace** [reply]
Там же игра в пьяницу? Навскидку первая половина выглядит достаточно просто.

**Reply by aragaer** [reply]
Потратил целый день, но все-таки переписал ее через State и тайпклассы.

**Reply by astynax** [reply]
Первая половина простая. Во второй мутная формулировка и дебаг с пристальным вглядыванием в пример.

**Comment by aragaer** [comment]
По второй части day23 - тупо в лоб c iterate' долго получается. В прошлом году задачка на тасовку карт решалась через математику. Тут что-то похожее?

**Reply by astynax** [reply]
Я тоже попробовал в лоб, и да - долго :)

**Reply by astynax** [reply]
На Reddit пишут, что реализации на мутабельных linked lists отрабатывают за единицы секунд даже 10М шагов %)

**Reply by aragaer** [reply]
Есть какая-нибудь готовая библиотека? Ну или краткое руководстов, как переписать с обычных списков на такие шустрые?

**Reply by ShrykeWindgrace** [reply]
Я условие только читал, но выглядит всё именно так.

**Comment by astynax** [comment]
Обычные тоже шустрые, но не в таком сценарии использования. Двусвязные
мутабельные списки можно сделать на Foreign Pointers, наверное

**Reply by aragaer** [reply]
Может есть готовая реализация? Гугл выдал мне какую-то статью по mergesort using mutable arrays, но там ссылка ведет на протухший hpaste.

**Comment by astynax** [comment]
Не нужен тут array, в том то и дело! Двусвязный список позволяет вырезать и
вставлять куски за константное время и реализовывать замкнутое кольцо,
которое можно прокручивать в обе стороны

**Reply by aragaer** [reply]
Ну там требуется поиск элемента по значению - это линейное время. Но ускорить разрезание и переклеивание очень хочется.

Короче, если это надо писать самому, то я на такое не способен. Если есть готовая библиотека с такими структурами или хотя бы пример реализации, то было бы очень кстати.

**Comment by astynax** [comment]
Линейный поиск дешевле, чем односвязный список перекладывать из стопки в
стопку

**Reply by aragaer** [reply]
Сделал мутабельный вектор, смотрю скорость...

**Reply by aragaer** [reply]
На мутабельных векторах 2 минуты на 1000 итераций. Все-таки нужны мутабельные списки.

**Reply by shnifer** [reply]
У меня на Data.Sequence получилось 7.8s на 10000 итераций, что быстрее чем на Data.Vector.Unboxed (13s), но всё равно ждать два часа и верить что скорость не сильно деградирует грустно

**Reply by aragaer** [reply]
Сделал совсем жесть - на базе Vector.Primitive.Mutable сделал эмуляцию связного списка (значение по индексу - номер следующей ячейки). 10 млн итераций за 0.9 секунды.

**Reply by aragaer** [reply]
Итак, достаточно быстро и шустро можно все сделать, если действительно иметь закольцованный single-linked list. А чтобы его иметь достаточно иметь cons-пары. А дальше я вспомнил старый прием, который еще в бейсике использовался - вместо "указателей" используются индексы в некотором массиве. Таким образом нам нужны пары вида "ключ-индекс", а их уже легко сделать даже той же Data.Map. Это может быть будет работать не за секунду, но каждая итерация превращается в 5 чтений и 4 записи, каждое из которых делается за константу.

**Reply by shnifer** [reply]
На мутабельном массиве связей и правда получается за единицы секунд! Но выглядит это с непривычки так, что Си кажется изящным )

**Reply by astynax** [reply]
Я сделал на `Vector.Primitive.Mutable`, но всё равно считал минут десять :) Зато код не слишком страшный вышел. Может быть созрею когда-нить попрофилировать и найти, где тормозит.

**Reply by astynax** [reply]
Но у меня двусвязный список. М.б. можно будет и отбросить вторую связь, вроде я в одну сторону хожу всё время...

**Comment by astynax** [comment]
Чёт 24 и 25 сделаны "на отвяжись". В итоге у меня `46*` из требуемых `49*`. При 49 задачах. Лучше бы позволили любую одну не сделать...
